api/upload.go

```diff
@@ -0,0 +1,87 @@
+package api
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/SkynetLabs/skynet-accounts/database"
+	"github.com/julienschmidt/httprouter"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+type (
+	// UploaderInfo gives information about a user who created an upload.
+	UploaderInfo struct {
+		UserID   primitive.ObjectID
+		Email    string
+		Sub      string
+		StripeID string
+	}
+
+	// UploadInfo gives information about a given upload.
+	UploadInfo struct {
+		Skylink    string
+		UploaderIP string
+		UploadedAt time.Time
+		UploaderInfo
+	}
+)
+
+// uploadInfoGET returns detailed information about all uploads of a given
+// skylink.
+func (api *API) uploadInfoGET(_ *database.User, w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	skylink := ps.ByName("skylink")
+	if !database.ValidSkylinkHash(skylink) {
+		api.WriteError(w, database.ErrInvalidSkylink, http.StatusBadRequest)
+		return
+	}
+
+	ctx := req.Context()
+	sl, err := api.staticDB.Skylink(ctx, skylink)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to get skylink"), http.StatusInternalServerError)
+		return
+	}
+	// Get all uploads of this skylink.
+	ups, err := api.staticDB.UploadsBySkylinkID(ctx, sl.ID)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to get uploads"), http.StatusInternalServerError)
+		return
+	}
+	// Get a set of all users that created these uploads.
+	uploaderSet := make(map[primitive.ObjectID]interface{})
+	for _, up := range ups {
+		uploaderSet[up.UserID] = struct{}{}
+	}
+	// Fet those users' data.
+	uploaders := make(map[primitive.ObjectID]database.User)
+	for upID := range uploaderSet {
+		u, err := api.staticDB.UserByID(ctx, upID)
+		if errors.Contains(err, database.ErrUserNotFound) {
+			continue
+		}
+		if err != nil {
+			api.WriteError(w, errors.AddContext(err, "failed to get uploader"), http.StatusInternalServerError)
+			return
+		}
+		uploaders[u.ID] = *u
+	}
+	// Create the final list of hydrated uploads.
+	upsHydrated := make([]UploadInfo, len(ups))
+	for i, up := range ups {
```

Comment:
hm so this means that if I pass you 2 skylinks, and you can't find any info for the first and you find 1 upload for the second, I only get one return element, I was kind of hoping the accounts API would return a shell object that only has the skylink

thinking about it more though, I'm not sure what's the cleanest design, I can obviously do this on my side as well